### PR TITLE
[SPIRV] Add OpAccessChain instruction support

### LINKED ISSUE #66107: [SPIRV] All SPIR-V pointer types are i8* instead of the proper type
This is a known issue for the SPIR-V backend, creating this to track the status.

Before opaque pointers, the LLVM types were lowered directly to their SPIR-V equivalent.
Now that we have opaque pointers, we need to retrieve the SPIR-V type from the typed operations like load/store.
The current backend can retrieve some types, but for pointers, it falls back on `i8*`. This is probably because OpenCL can do pointer-casting.

For graphical SPIR-V, we are not allowed to do pointer casts, so we need to figure out the correct pointer type. LLVM-SPIRV translator already has such logic (type scavenger), so we'd need to reproduce this logic in this backend.

### PR DESCRIPTION
**Pull request by Keenuts**
This commit adds 2 new instructions in the selector:
 - OpAccessChain
 - OpInBoundsAccessChain.

The choice between the two relies on the `inbounds` marker.

Those instruction are not used for OpenCL, to maintain the same
behavior as previously. They are only added when building for logical
SPIR-V, as it doesn't support the pointer equivalent.

Because logical SPIR-V doesn't support pointer cast either, the
assign_ptr_type intrinsic need to be generated so OpAccessChain gets
lowered with the correct pointer type, instead of i8*.

Fixes #66107

### CODE REVIEW

**Comment by Keenuts** [comment]
This PR depends on https://reviews.llvm.org/D156049?id=556194
@sudonatalie cannot add you as reviewer for now, but FYI.

File: llvm/test/CodeGen/SPIRV/logical-struct-access.ll [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  %6 = getelementptr inbounds %B, ptr %0, i32 0, i32 2
+; CHECK: {{%[0-9]+}} = OpInBoundsAccessChain [[ptr_A]] [[tmp]] [[uint_2]]
+
+  %7 = getelementptr %B, ptr %0, i32 0, i32 2, i32 1
+; CHECK: {{%[0-9]+}} = OpAccessChain [[ptr_uint]] [[tmp]] [[uint_2]] [[uint_1]]
+  %8 = getelementptr inbounds %B, ptr %0, i32 0, i32 2, i32 1
+; CHECK: {{%[0-9]+}} = OpInBoundsAccessChain [[ptr_uint]] [[tmp]] [[uint_2]] [[uint_1]]
+
+  %9 = getelementptr %B, ptr %0, i32 0, i32 2
+	%10 = getelementptr %A, ptr %9, i32 0, i32 1
```
Comment (sudonatalie) [inline llvm/test/CodeGen/SPIRV/logical-struct-access.ll:52]: Least important possible nit, but there's an extra indent here

**Review by sudonatalie — APPROVED** [review]

File: llvm/test/CodeGen/SPIRV/logical-struct-access.ll (hunk 1 above)
Comment (Keenuts) [reply llvm/test/CodeGen/SPIRV/logical-struct-access.ll]: Arf this was a tab. Thanks!

File: llvm/test/CodeGen/SPIRV/logical-access-chain.ll [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+; CHECK: [[var:%[0-9]+]] = OpVariable [[ptr_uint2]] Function
+
+  %1 = getelementptr <2 x i32>, ptr %0, i32 0, i32 1
+; CHECK: {{%[0-9]+}} = OpAccessChain [[ptr_uint]] [[var]] [[uint_1]]
+
+  ret void
+}
+
+attributes #1 = { "hlsl.numthreads"="4,8,16" "hlsl.shader"="compute" }
+
```
Comment (michalpaszkowski) [inline llvm/test/CodeGen/SPIRV/logical-access-chain.ll:21]: Please remove the trailing new line (single new line at the end of file).

**Review by michalpaszkowski — APPROVED** [review]
Thank you @Keenuts for the patch and including useful comments in the selector :) Those help a lot later on. The patch LGTM! Please remove the double blank lines in one LIT test.

**Comment by Keenuts** [comment]
Rebased on main. Running tests now, will merge if nothing broke.

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp [hunk 3]
```diff
@@ -1299,18 +1299,28 @@ bool SPIRVInstructionSelector::selectExtractElt(Register ResVReg,
 bool SPIRVInstructionSelector::selectGEP(Register ResVReg,
                                          const SPIRVType *ResType,
                                          MachineInstr &I) const {
-  // In general we should also support OpAccessChain instrs here (i.e. not
-  // PtrAccessChain) but SPIRV-LLVM Translator doesn't emit them at all and so
-  // do we to stay compliant with its test and more importantly consumers.
-  unsigned Opcode = I.getOperand(2).getImm() ? SPIRV::OpInBoundsPtrAccessChain
-                                             : SPIRV::OpPtrAccessChain;
+  const bool isGEPInBounds = I.getOperand(2).getImm();
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:1302]: Ah, that timing is a bit unfortunate 🙃 Sent a PR.

[the remark this answers is not in the document]
Comment (iliya-diyachkov) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:1302]: Thank you, approved. Please add me to reviewers next time so I have time to look at the patch .

**Pull request merged**